google-cloud-storage/src/test/java/com/google/cloud/storage/conformance/retry/CtxFunctions.java

```diff
@@ -0,0 +1,124 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage.conformance.retry;
+
+import static com.google.common.collect.Sets.newHashSet;
+
+import com.google.cloud.conformance.storage.v1.Resource;
+import com.google.cloud.storage.Acl;
+import com.google.cloud.storage.Acl.Role;
+import com.google.cloud.storage.Acl.User;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Bucket;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.conformance.retry.Functions.CtxFunction;
+import com.google.common.base.Joiner;
+import java.util.HashSet;
+
+final class CtxFunctions {
+
+  private static final class Util {
+    private static final CtxFunction blobIdAndBlobInfo =
+        (ctx, c) -> ctx.map(state -> state.with(BlobInfo.newBuilder(state.getBlobId()).build()));
+  }
+
+  static final class Local {
+    static final CtxFunction blobCopy =
+        (ctx, c) -> ctx.map(s -> s.withCopyDest(BlobId.of(c.getBucketName2(), c.getObjectName())));
+
+    static final CtxFunction bucketInfo =
+        (ctx, c) -> ctx.map(s -> s.with(BucketInfo.of(c.getBucketName())));
+    static final CtxFunction blobIdWithoutGeneration =
+        (ctx, c) -> ctx.map(s -> s.with(BlobId.of(c.getBucketName(), c.getObjectName())));
+    static final CtxFunction blobIdWithGenerationZero =
+        (ctx, c) -> ctx.map(s -> s.with(BlobId.of(c.getBucketName(), c.getObjectName(), 0L)));
+    static final CtxFunction blobInfoWithoutGeneration =
+        blobIdWithoutGeneration.andThen(Util.blobIdAndBlobInfo);
+    static final CtxFunction blobInfoWithGenerationZero =
+        blobIdWithGenerationZero.andThen(Util.blobIdAndBlobInfo);
+  }
+
+  static final class Rpc {
+    static final CtxFunction bucket =
+        (ctx, c) ->
+            ctx.map(state -> state.with(ctx.getStorage().get(state.getBucketInfo().getName())));
+    static final CtxFunction blobWithGeneration =
+        (ctx, c) -> ctx.map(state -> state.with(ctx.getStorage().get(state.getBlobId())));
+    static final CtxFunction createEmptyBlob =
+        (ctx, c) -> ctx.map(state -> state.with(ctx.getStorage().create(state.getBlobInfo())));
+  }
+
+  static final class ResourceSetup {
+    private static final CtxFunction bucket =
+        (ctx, c) -> {
+          BucketInfo bucketInfo = BucketInfo.newBuilder(c.getBucketName()).build();
+          Bucket resolvedBucket = ctx.getStorage().create(bucketInfo);
+          return ctx.map(s -> s.with(resolvedBucket));
+        };
+    private static final CtxFunction object =
+        (ctx, c) -> {
+          BlobInfo blobInfo =
+              BlobInfo.newBuilder(ctx.getState().getBucket().getName(), c.getObjectName()).build();
+          Blob resolvedBlob = ctx.getStorage().create(blobInfo);
+          return ctx.map(s -> s.with(resolvedBlob));
+        };
```

Comment:
Assuming HMAC key and Notification will go here as well eventually?

Reply:
Yup, I actually added it this morning.

Reply:
Notification is turning out to be a bit of a unique case in the java library as it's not part of the `Storage` api, only the `StorageRPC` api which doesn't have any retry at all in that layer. So it'll take some thinking to figure out how we're going to manage that.